Tasks/Npm/npm.ts

```diff
@@ -0,0 +1,273 @@
+import * as path from "path";
+import * as url from "url";
+
+import * as tl from "vsts-task-lib/task";
+
+import { NpmCommand, NpmTaskInput, RegistryLocation } from "./constants";
+import { GetRegistries, NormalizeRegistry } from "./npmrcparser";
+import { NpmToolRunner } from "./npmtoolrunner";
+
+interface INpmRegistry {
+    url: string;
+    auth: string;
+    authOnly: boolean;
+}
+
+class NpmRegistry implements INpmRegistry {
+    public url: string;
+    public auth: string;
+    public authOnly: boolean;
+
+    constructor(url: string, auth: string, authOnly?: boolean) {
+        this.url = url;
+        this.auth = auth;
+        this.authOnly = authOnly || false;
+
+        tl.debug(authOnly ? tl.loc("NpmRegistryCreatedAuth", url) : tl.loc("NpmRegistryCreated", url));
+    }
+
+    public static FromServiceEndpoint(endpointId: string, authOnly?: boolean): NpmRegistry {
+        let email: string;
+        let username: string;
+        let password: string;
+        let endpointAuth: tl.EndpointAuthorization;
+        let url: string;
+        try {
+            endpointAuth = tl.getEndpointAuthorization(endpointId, false);
+        }
+        catch (exception) {
+            throw new Error(tl.loc("ServiceEndpointNotDefined"));
+        }
+
+        try {
+            url = NormalizeRegistry(tl.getEndpointUrl(endpointId, false));
+        }
+        catch (exception) {
+            throw new Error(tl.loc("ServiceEndpointUrlNotDefined"));
+        }
+
+        let trimmedUrl = NpmRegistry._trimRegistryUrl(url);
+
+        switch (endpointAuth.scheme) {
+            case "UsernamePassword":
+                username = endpointAuth.parameters["username"];
+                password = endpointAuth.parameters["password"];
+                email = username; // npm needs an email to be set in order to publish, this is ignored on npmjs
+                break;
+            case "Token":
+                email = "VssEmail";
+                username = "VssToken";
+                password = endpointAuth.parameters["apitoken"];
+                break;
+        }
+
+        let auth = `${trimmedUrl}:username=${username}
+                    ${trimmedUrl}:_password=${new Buffer(password).toString("base64")}
+                    ${trimmedUrl}:email=${email}
+                    ${trimmedUrl}:always-auth=true`;
+
+        return new NpmRegistry(url, auth, authOnly);
+    }
+
+    public static FromFeedId(feedId: string, authOnly?: boolean): NpmRegistry {
+        let collectionUrl = tl.getVariable("System.TeamFoundationCollectionUri");
+        collectionUrl = NormalizeRegistry(collectionUrl);
+
+        let url = `${collectionUrl}_packaging/${feedId}/npm/registry/`;
```

Comment:
Look at [this](https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/NuGetRestore/nugetinstaller.ts#L247-L275).